### PR TITLE
(1) Add UnintrusiveBindingPathProvider

### LINKED ISSUE #4170: Add unintrusive path provider
### Description

Implement `UnintrusiveBindingPathProvider`. Returns a solution-specific location under the standard SLVS user-specific folder
i.e. `$(APPDATA)\SonarLint for Visual Studio\\Bindings\\$(SolutionName)_$(SolutionDir.GetHashCode())\binding.config`

See #4168 for more details

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #4170

### CODE REVIEW

File: src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    /// <summary>
+    /// Return the path of solution's binding configuration file when in connected mode.
+    /// </summary>
+    internal interface IUnintrusiveBindingPathProvider
+    {
+        string Get();
+    }
+
+    [Export(typeof(IUnintrusiveBindingPathProvider))]
+    internal class UnintrusiveBindingPathProvider : IUnintrusiveBindingPathProvider
```
Comment (duncanp-sonar) [inline src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs:39]: This started as a copy of the old [ConnectedModeBindingPathProvider](https://github.com/SonarSource/sonarlint-visualstudio/blob/1f232ed6c1f7ee3d35b9af02222bac33553e306d/src/Integration/NewConnectedMode/ConnectedModeSolutionBindingPathProvider.cs#L32).

I tweaked it to be MEF exportable and explicitly fetch the VS service on the UI thread.

File: src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            IVsSolution slnService = null;
+            threadHandling.RunOnUIThreadSync(() => slnService = serviceProvider.GetService<SVsSolution, IVsSolution>());
+            solution = slnService;
+        }
+
+        public string Get()
+        {
+            // If there isn't an open solution the returned hresult will indicate an error
+            // and the returned solution name will be null. We'll just ignore the hresult.
+            solution.GetProperty((int)__VSPROPID.VSPROPID_SolutionFileName, out var fullSolutionName);
```
Comment (duncanp-sonar) [inline src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs:67]: There's a separate ticket to fix up the path in the targets file (#4194) once the location of all the files has been finalised (#4188).

File: src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using SonarLint.VisualStudio.Core;
+
+namespace SonarLint.VisualStudio.Integration.UnintrusiveBinding
+{
+    /// <summary>
+    /// Return the path of solution's binding configuration file when in connected mode.
+    /// </summary>
+    internal interface IUnintrusiveBindingPathProvider
```
Comment (duncanp-sonar) [inline src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs:33]: I've deliberately named the new interfaces and classes differently and put them in a separate folder to make the difference between the old and new code more obvious.
The goal in the longer term is to move most of the binding classes to the `ConnectedMode` assembly. For now I've put them in `Integration` along with the old code.

File: src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            IThreadHandling threadHandling)
+            : this(serviceProvider, threadHandling, EnvironmentVariableProvider.Instance)
+        {
+        }
+
+        internal /* for testing */ UnintrusiveBindingPathProvider(IServiceProvider serviceProvider,
+            IThreadHandling threadHandling,
+            IEnvironmentVariableProvider environmentVariables)
+        {
+            SLVSRootBindingFolder = Path.Combine(environmentVariables.GetSLVSAppDataRootPath(), "Bindings");
```
Comment (rita-gorokhod) [inline src/Integration/UnintrusiveBinding/UnintrusiveBindingPathProvider.cs:56]: I opened this https://github.com/SonarSource/sonarlint-visualstudio/issues/4207

**Pull request merged**